File: apps/shared/Sources/FerniShared/BetterThanHuman/MicroExpressions.swift
import SwiftUI
import Combine

// MARK: - Micro-Expression Engine
/// Manages subliminal emotional flashes lasting 40-150ms.
/// These are below conscious perception but affect how users FEEL about Ferni's authenticity.
///
/// From BETTER-THAN-HUMAN.md:
/// - Recognition: 80ms - User mentions familiar topic
/// - Concern Flash: 60ms - Before empathy kicks in
/// - Delight Flash: 100ms - User achieves something
/// - Warmth Pulse: 120ms - Connection moments
/// - Interest Flash: 70ms - Unexpected content

public class MicroExpressionEngine: ObservableObject {

    // MARK: - Published State

    @Published public private(set) var activeExpression: MicroExpressionType? = nil
    @Published public private(set) var expressionIntensity: CGFloat = 0

    // MARK: - Private State

    private var resetTask: DispatchWorkItem?

    // MARK: - Public API

    public init() {}

    /// Trigger a micro-expression
    /// - Parameter type: The type of micro-expression to display
    public func trigger(_ type: MicroExpressionType) {
        // Cancel any pending reset
        resetTask?.cancel()

        // Set expression immediately
        activeExpression = type
        expressionIntensity = type.intensity

        // Schedule reset after expression duration
        let task = DispatchWorkItem { [weak self] in
            self?.reset()
        }
        resetTask = task

        DispatchQueue.main.asyncAfter(deadline: .now() + type.duration, execute: task)
    }

    /// Reset to neutral state
    private func reset() {
        // Quick fade out (half the trigger duration)
        withAnimation(.easeOut(duration: (activeExpression?.duration ?? 0.08) / 2)) {
            expressionIntensity = 0
        }

        // Clear expression after fade
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.05) { [weak self] in
            self?.activeExpression = nil
        }
    }
}

// MARK: - Micro-Expression Types

public enum MicroExpressionType: String, Equatable, CaseIterable {
    // Recognition & Connection
    case recognition    // User mentions familiar topic
    case memorySpark    // Something triggers shared history (EASTER EGG)
    case insider        // Inside joke recognition - brief knowing look

    // Concern & Care
    case concern        // Before empathy kicks in
    case protective     // When sensing vulnerability

    // Interest & Engagement
    case interest       // Unexpected content
    case curiosityPeak  // Genuine interest flash
    case delight        // User achieves something

    // Understanding
    case epiphany       // "Aha!" moment
    case connection     // Mutual understanding flash

    // Warmth
    case warmth         // Connection moments
    case affection      // Brief caring expression

    /// Duration in seconds (subliminal: 40-150ms)
    /// ENFORCED: Never below 40ms or above 150ms
    public var duration: TimeInterval {
        switch self {
        case .recognition: return 0.08    // 80ms
        case .memorySpark: return 0.10    // 100ms - slightly longer for memory
        case .insider: return 0.09        // 90ms - knowing look
        case .concern: return 0.06        // 60ms - quick flash
        case .protective: return 0.07     // 70ms
        case .interest: return 0.07       // 70ms
        case .curiosityPeak: return 0.05  // 50ms - fastest flash
        case .delight: return 0.10        // 100ms - let joy be seen
        case .epiphany: return 0.06       // 60ms - snap realization
        case .connection: return 0.08     // 80ms
        case .warmth: return 0.12         // 120ms - warmest
        case .affection: return 0.08      // 80ms
        }
    }

    /// Visual intensity (0-1)
    public var intensity: CGFloat {
        switch self {
        case .recognition: return 0.4
        case .memorySpark: return 0.5     // Stronger for memory recognition
        case .insider: return 0.4
        case .concern: return 0.3
        case .protective: return 0.35
        case .interest: return 0.35
        case .curiosityPeak: return 0.35
        case .delight: return 0.6         // Brightest for joy
        case .epiphany: return 0.6
        case .connection: return 0.45
        case .warmth: return 0.5
        case .affection: return 0.4
        }
    }

    /// Effect on the avatar soul/glow
    public var soulEffect: MicroSoulEffect {
        switch self {
        case .recognition:
            return MicroSoulEffect(warmthOpacity: 0.3, sparkOpacity: 0.4, shimmerBoost: 0)
        case .memorySpark:
            return MicroSoulEffect(warmthOpacity: 0.4, sparkOpacity: 0.6, shimmerBoost: 0.15)
        case .insider:
            return MicroSoulEffect(warmthOpacity: 0.35, sparkOpacity: 0.3, shimmerBoost: 0.1)
        case .concern:
            return MicroSoulEffect(warmthOpacity: 0.2, sparkOpacity: 0, shimmerBoost: -0.1)
        case .protective:
            return MicroSoulEffect(warmthOpacity: 0.25, sparkOpacity: 0, shimmerBoost: 0)
        case .interest:
            return MicroSoulEffect(warmthOpacity: 0.1, sparkOpacity: 0.3, shimmerBoost: 0.15)
        case .curiosityPeak:
            return MicroSoulEffect(warmthOpacity: 0.15, sparkOpacity: 0.35, shimmerBoost: 0.2)
        case .delight:
            return MicroSoulEffect(warmthOpacity: 0.4, sparkOpacity: 0.6, shimmerBoost: 0.2)
        case .epiphany:
            return MicroSoulEffect(warmthOpacity: 0.3, sparkOpacity: 0.7, shimmerBoost: 0.25)
        case .connection:
            return MicroSoulEffect(warmthOpacity: 0.45, sparkOpacity: 0.3, shimmerBoost: 0.1)
        case .warmth:
            return MicroSoulEffect(warmthOpacity: 0.5, sparkOpacity: 0.2, shimmerBoost: 0.1)
        case .affection:
            return MicroSoulEffect(warmthOpacity: 0.4, sparkOpacity: 0.25, shimmerBoost: 0.05)
        }
    }

    /// Glow color for this expression
    public var glowColor: Color {
        switch self {
        case .recognition, .memorySpark, .insider, .warmth, .affection, .connection:
            return Color(hexString: "c4a265")  // Gold - recognition/warmth
        case .concern, .protective:
            return Color(hexString: "9a8a82")  // Muted warm - concern
        case .curiosityPeak, .interest:
            return Color(hexString: "7a9a7a")  // Soft green - curiosity
        case .delight, .epiphany:
            return .white  // Bright - joy
        }
    }

    /// Animation easing for this expression
    public var easing: Animation {
        return .easeOut(duration: duration)
    }
}

// MARK: - Soul Effect Values

/// Visual effect values to apply to the avatar's soul/glow layers
public struct MicroSoulEffect: Equatable {
    /// Warmth glow opacity boost (0-1)
    public var warmthOpacity: CGFloat

    /// Memory spark opacity (0-1)
    public var sparkOpacity: CGFloat

    /// Shimmer intensity boost (-0.2 to 0.3)
    public var shimmerBoost: CGFloat

    public init(warmthOpacity: CGFloat = 0, sparkOpacity: CGFloat = 0, shimmerBoost: CGFloat = 0) {
        self.warmthOpacity = warmthOpacity
        self.sparkOpacity = sparkOpacity
        self.shimmerBoost = shimmerBoost
    }
}

File: apps/android-native/app/src/main/java/com/ferni/voice/betterthanuman/MicroExpressions.kt
package com.ferni.voice.betterthanuman

import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Micro-Expression Engine
 *
 * Manages subliminal emotional flashes lasting 40-150ms.
 * These are below conscious perception but affect how users FEEL about Ferni's authenticity.
 *
 * From BETTER-THAN-HUMAN.md:
 * - Recognition: 80ms - User mentions familiar topic
 * - Concern Flash: 60ms - Before empathy kicks in
 * - Delight Flash: 100ms - User achieves something
 * - Warmth Pulse: 120ms - Connection moments
 * - Interest Flash: 70ms - Unexpected content
 */
class MicroExpressionEngine(
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Main)
) {

    // MARK: - State

    private val _activeExpression = MutableStateFlow<MicroExpressionType?>(null)
    val activeExpression: StateFlow<MicroExpressionType?> = _activeExpression.asStateFlow()

    private val _expressionIntensity = MutableStateFlow(0f)
    val expressionIntensity: StateFlow<Float> = _expressionIntensity.asStateFlow()

    private var resetJob: Job? = null

    // MARK: - Public API

    /**
     * Trigger a micro-expression.
     * @param type The type of micro-expression to display
     */
    fun trigger(type: MicroExpressionType) {
        // Cancel any pending reset
        resetJob?.cancel()

        // Set expression immediately
        _activeExpression.value = type
        _expressionIntensity.value = type.intensity

        // Schedule reset after expression duration
        resetJob = scope.launch {
            delay(type.durationMs)
            reset()
        }
    }

    /**
     * Reset to neutral state
     */
    private suspend fun reset() {
        // Quick fade out (half the trigger duration)
        val fadeOutDuration = (_activeExpression.value?.durationMs ?: 80L) / 2

        // Animate intensity to 0
        val startIntensity = _expressionIntensity.value
        val steps = 10
        val stepDelay = fadeOutDuration / steps

        for (i in steps downTo 0) {
            _expressionIntensity.value = startIntensity * (i.toFloat() / steps)
            delay(stepDelay)
        }

        // Clear expression after fade
        delay(50)
        _activeExpression.value = null
    }

    /**
     * Clear any active expression immediately
     */
    fun clear() {
        resetJob?.cancel()
        _activeExpression.value = null
        _expressionIntensity.value = 0f
    }
}

/**
 * Micro-Expression Types
 *
 * Each type has specific timing (subliminal: 40-150ms) and visual effects
 * that make Ferni feel emotionally present without being consciously noticed.
 */
enum class MicroExpressionType(
    /** Duration in milliseconds (subliminal: 40-150ms) */
    val durationMs: Long,
    /** Visual intensity (0-1) */
    val intensity: Float,
    /** Effect on the avatar soul/glow */
    val soulEffect: MicroSoulEffect
) {
    /**
     * User mentions familiar topic - 80ms
     */
    RECOGNITION(
        durationMs = 80,
        intensity = 0.4f,
        soulEffect = MicroSoulEffect(warmthOpacity = 0.3f, sparkOpacity = 0.4f, shimmerBoost = 0f)
    ),

    /**
     * Before empathy kicks in - 60ms
     */
    CONCERN(
        durationMs = 60,
        intensity = 0.3f,
        soulEffect = MicroSoulEffect(warmthOpacity = 0.2f, sparkOpacity = 0f, shimmerBoost = -0.1f)
    ),

    /**
     * User achieves something - 100ms
     */
    DELIGHT(
        durationMs = 100,
        intensity = 0.6f,
        soulEffect = MicroSoulEffect(warmthOpacity = 0.4f, sparkOpacity = 0.6f, shimmerBoost = 0.2f)
    ),

    /**
     * Connection moments - 120ms
     */
    WARMTH(
        durationMs = 120,
        intensity = 0.5f,
        soulEffect = MicroSoulEffect(warmthOpacity = 0.5f, sparkOpacity = 0.2f, shimmerBoost = 0.1f)
    ),

    /**
     * Unexpected content - 70ms
     */
    INTEREST(
        durationMs = 70,
        intensity = 0.35f,
        soulEffect = MicroSoulEffect(warmthOpacity = 0.1f, sparkOpacity = 0.3f, shimmerBoost = 0.15f)
    );

    /**
     * Duration in seconds for animation compatibility
     */
    val durationSeconds: Float
        get() = durationMs / 1000f
}

/**
 * Visual effect values to apply to the avatar's soul/glow layers.
 */
data class MicroSoulEffect(
    /** Warmth glow opacity boost (0-1) */
    val warmthOpacity: Float = 0f,

    /** Memory spark opacity (0-1) */
    val sparkOpacity: Float = 0f,

    /** Shimmer intensity boost (-0.2 to 0.3) */
    val shimmerBoost: Float = 0f
)
